text/0000-code-literals.md

```diff
@@ -0,0 +1,517 @@
+        of lines can\
+        be ignored\
+        "
+        ```
+
+        This approach still suffers from the need to escape special
+        characters. The backslashes at the end of every line are
+        tedious to write, and are problematic if whitespace is
+        meaningful within the code. For example, if python code
+        was being embedded, then the indentation would be lost.
+        Finally, although `rustfmt` could in principle reformat
+        these strings, in practice doing so in a reasonable way
+        is complicated and so this has never been enabled by default.
+    
+    3.  Raw string literals, eg. `r#"I can use "s!"#`
+
+        This solves the problem of special characters, but suffers
+        from the same inability to be reformatted, and the trick
+        of using an `\` at the end of each line cannot be applied
+        because escape characters are not recognised.
+
+  - The existing string literals introduce extra unwanted whitespace
+    into the literal value. Even if that extra whitespace does not
+    semantically affect the nested code, it results in ugly output
+    if the code is ever logged (such as might happen when logging
+    SQL query executions).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A modifier `h` (for
+[Here document](https://en.wikipedia.org/wiki/Here_document))
+may be added to a string literal prefix to change how the
+string is interpreted by the compiler. The effect of the `h`
+modifier causes all indentation to be relative to the
+closing quote:
+
+```rust
+    let code = h"
+        This is a code string literal.
+
+        I can use escape sequences like \n since the `h`
+        prefix was added to a normal string literal
+
+            Indentation is preserved *relative* to the indentation level
+            of the terminating quote.
+
+    It is an error for a line to have negative indentation (ie. be
+    indented less than the final quote) unless
+    the line is empty.
+        ";
+```
+
+`rustfmt` will automatically adjust the indentation of the code string
+literal as a whole to match the surrounding context, but will never
+change the relative indentation within such a literal.
+
+The `h` modifier will often be combined with raw string literals to
+embed sections of code such as SQL:
+
+```rust
+    let code = hr#"
+        This is also a code string literal
+
+        I can use special characters like "" and \ freely.
+
+            Indentation is still *relative* to the indentation level
+            of the terminating quote.
+        "#;
+```
+
+For completeness, the `h` modifier may also be combined with byte
+and raw byte string literals, eg. `hb"` and `hbr#"`.
+
+Anything directly after the opening quote is not considered
+part of the string literal. It may be used as a language hint or
+processed by macros (similar to the treatment of doc comments).
+
+```rust
+let sql = hr#"sql
+    SELECT * FROM table;
+    "#;
+```
```

Comment:
I do not believe placing metadata regarding the string should be accepted, as it has many negative impacts. In particular, there is an isomorphism between strings written using `r#""#` and strings written using `""`, currently, that as far as I know is complete. This proposal would create a surjective function: there would be string literals written using
```rust
h"
"
```
which have no mirror image using the other syntactic forms for string literals. This causes great amounts of confusion for:
- Lexing
- Parsing
- Code generation

And the very purpose of this language hint is for the service of syntax highlighters and the like, which are very likely going to be written in a language that may have no easy access to simply running syn or tree-sitter or whatever, and may instead be bashed together out of JavaScript and regexes.

Reply:
>  In particular, there is an isomorphism between strings written using r#""# and strings written using "" (and without STRING_CONTINUE, i.e. 0x5C 0x0A), currently, that as far as I know is complete.

Not sure what you're getting at here. There are already many ways to get the same "literal value" using different "encodings" of the same literal. For example, tabs could be encoded with `\t` or an actual tab character.

> This proposal would create a surjective function: there would be string literals written using
> 
> h"languagetag
> "
> 
> which have no mirror image using the other syntactic forms for string literals.
> 
> 
> This causes great amounts of confusion for:
> 
>     Lexing
>     Parsing
>     Code generation

This is going to need some more justification.

First of all, the language hint is purely a syntactic feature, it doesn't change the "value" of a string literal, so in terms of "values" (which if we're using set theoretic terms, is the most plausible thing to talk about but you haven't actually defined that...) there is the same amount of isomorphism between code string literals and raw string literals as there was between raw string literals and string literals (modulo indentation being relative, which is the entire point of the proposal).

Secondly, I flat out don't believe that this does introduce significant complexity in those areas. The compiler/tooling is already capable of dealing with string literals and raw string literals. This RFC doesn't change the basic rules for when a string begins/ends - the parsing rules are identical to the corresponding non-code-literal form. The only change is to how the content within the literal is converted into a value for use by the program.